drivers/video/ov5640.c

```diff
@@ -916,19 +903,88 @@ static int ov5640_enum_frmival(const struct device *dev, enum video_endpoint_id
 	return 0;
 }
 
+static int ov5640_init_controls(const struct device *dev)
+{
+	int ret;
+	struct ov5640_data *drv_data = dev->data;
+	struct ov5640_ctrls *ctrls = &drv_data->ctrls;
+
+	video_ctrl_handler_init(&ctrls->handler, dev);
+
+	ret = video_ctrl_init(&ctrls->brightness, &ctrls->handler, VIDEO_CID_BRIGHTNESS, -15, 15, 1,
+			      0);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->contrast, &ctrls->handler, VIDEO_CID_CONTRAST, 0, 255, 1, 0);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->hue, &ctrls->handler, VIDEO_CID_HUE, 0, 359, 1, 0);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->saturation, &ctrls->handler, VIDEO_CID_SATURATION, 0, 255, 1,
+			      64);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->hflip, &ctrls->handler, VIDEO_CID_HFLIP, 0, 1, 1, 0);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->vflip, &ctrls->handler, VIDEO_CID_VFLIP, 0, 1, 1, 0);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->light_freq, &ctrls->handler, VIDEO_CID_POWER_LINE_FREQUENCY,
+			      VIDEO_CID_POWER_LINE_FREQUENCY_DISABLED,
+			      VIDEO_CID_POWER_LINE_FREQUENCY_AUTO, 1,
+			      VIDEO_CID_POWER_LINE_FREQUENCY_50HZ);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_ctrl_init(&ctrls->test_pattern, &ctrls->handler, VIDEO_CID_TEST_PATTERN, 0,
+			      ARRAY_SIZE(test_pattern_val) - 1, 1, 0);
+	if (ret) {
+		return ret;
+	}
+
+	return video_ctrl_init(
+		&ctrls->pixel_rate, &ctrls->handler, VIDEO_CID_PIXEL_RATE,
+		mipi_vga_frmrate_params[0].pixelrate,
+		mipi_hd_frmrate_params[ARRAY_SIZE(mipi_hd_frmrate_params) - 1].pixelrate, 1,
+		drv_data->cur_pixrate);
```

Comment:
Is there any chance that a Zephyr video device dynamically allocate controls? If not, this table of control fields could become a static array defined at compile time, in the style of struct format caps.

This could be done by putting [these fields](https://github.com/zephyrproject-rtos/zephyr/pull/82158/files#diff-981325ebbbf81f0cf401399c4f79e441ef5143e0662cafa3a87077e54b8a1c5aR145-R153) in an array and initializing them:
```c
const size_t npat = ARRAY_SIZE(test_pattern_val) - 1;

static struct video_ctrl ctrls[] = {
        /* handler,             CID,                    min,    max,    step,   def */
        {&ctrls->handler,       VIDEO_CID_BRIGHTNESS,   -15,    15,     1,      0},
        {&ctrls->handler,       VIDEO_CID_CONTRAST,     0,      255,    1,      0},
        {&ctrls->handler,       VIDEO_CID_HUE,          0,      359,    1,      0},
        {&ctrls->handler,       VIDEO_CID_SATURATION,   0,      255,    1,      64},
        {&ctrls->handler,       VIDEO_CID_HFLIP,        0,      1,      1,      0},
        {&ctrls->handler,       VIDEO_CID_VFLIP,        0,      1,      1,      0},
        {&ctrls->handler,       VIDEO_CID_TEST_PATTERN, 0,      npat,   1,      0);
        {&ctrls->handler,       VIDEO_CID_POWER_LINE_FREQUENCY,
                VIDEO_CID_POWER_LINE_FREQUENCY_DISABLED,
                VIDEO_CID_POWER_LINE_FREQUENCY_AUTO,
                1,
                VIDEO_CID_POWER_LINE_FREQUENCY_50HZ},
        {&ctrls->handler,       VIDEO_CID_PIXEL_RATE,
                mipi_vga_frmrate_params[0].pixelrate,
                mipi_hd_frmrate_params[ARRAY_SIZE(mipi_hd_frmrate_params) - 1].pixelrate,
                1,
                drv_data->cur_pixrate);
        {0},
};
```

Which does not prevent filling these fields dynamically at init time, or updating some fields later-on.

Reply:
With a table, implementing VIDEO_CID_PIXEL_RATE would become very ugly, that is true.

If someone wants to use tables (i.e. generate a table from vendor data), they can still use tables and loop over it themselves.

This means using a table is not a good idea, and even if it was, `video_ctrl_init()` would still be the ideal API. :)